### PR TITLE
Sprint 1

### PR DESCRIPTION
**Pull request by nntai**
## Summary
* Short summary of the task, what have been done etc
* Please include screenshots whenever possible (important).

## Ref
Solve #1 #2 #3 #4 #5 

## Lasting Changes (Technical)
(please list down: code changes/things that have wide-effect; new libraries/functions added that can be used by others; examples below)

* (Off Eslint rule 'vue/multi-word-component-names')
* (Add package 'xyz')

## Checklist

Please check directly on the box once each of these are done

- [ ] Review code before submit
- [ ] Update CHANGELOG.md
- [ ] Tests (integration test/unit test)
- [ ] Documentation & Demo (if necessary)

### CODE REVIEW

File: src/CommandPalette.vue [hunk 1]
```diff
@@ -1,20 +1,62 @@
 <template>
-  <div class="command-palette-wrapper">
-    <cmp-input class="cmp-input" />
-    <cmp-list class="cmp-list" />
+  <div v-show="customerModalController.isModal.value" class="command-pallete-place">
+    <div class="command-pallete-wrapper" v-click-outside="customerModalController.closeModal">
+      <cmp-input class="cmp-input" v-bind:inputText="textInputController" />
+      <cmp-list class="cmp-list" v-bind:customerCommands="commandsController.customerCommands.value" />
```
Comment (nntai) [inline src/CommandPalette.vue:5]:
```suggestion
      <cmp-list class="cmp-list" :customer-commands="commandsController.customerCommands.value" />
```

File: src/CommandPalette.vue [hunk 2]
```diff
@@ -1,20 +1,62 @@
...
+
+import customerInputController from "./controllers/customerInputController";
+
+import customerCommandController from "./controllers/customerCommandController";
+
+import modalController from "./controllers/modalController";
+
+import { onMounted, computed, watch, ref } from "vue";
+
+import {vClickOutside} from "./directives/vClickOutside";
```
Comment (nntai) [inline src/CommandPalette.vue:23]:
```suggestion
import { vClickOutside } from "./directives/vClickOutside";
```

File: src/Demo.vue [hunk 3]
```diff
@@ -1,10 +1,13 @@
 <template>
-  <command-palette class="command-palette"/>
+  <command-pallete class="command-pallete" v-bind:customerCommands="[{commandName: 'place', commandKey: 'm', commandAction: () => {placeIn();}}]" />
```
Comment (nntai) [inline src/Demo.vue:2]: `customerCommands` should be a computed

File: src/components/CmpInput.vue [hunk 4]
```diff
@@ -1,10 +1,26 @@
...
+  
+  const props = defineProps({
+    inputText: {
+      type: Object,
+      default: function(placeProps) {
+        return {
+          customerInput: ref("")
+        };
+      }
+    }
```
Comment (nntai) [inline src/components/CmpInput.vue:19]:
```suggestion
    inputText: {
      type: String,
      default: ''
      }
    }
```

File: src/CommandPalette.vue [hunk 5]
```diff
@@ -1,20 +1,62 @@
 <template>
-  <div class="command-palette-wrapper">
-    <cmp-input class="cmp-input" />
-    <cmp-list class="cmp-list" />
+  <div v-show="customerModalController.isModal.value" class="command-pallete-place">
+    <div class="command-pallete-wrapper" v-click-outside="customerModalController.closeModal">
+      <cmp-input class="cmp-input" v-bind:inputText="textInputController" />
```
Comment (nntai) [inline src/CommandPalette.vue:4]:
```javascript
const textInputController = customerInputController();
```
=> 
```js
const { value: inputText } = customerInputController();
```

```suggestion
      <cmp-input class="cmp-input" v-bind:inputText="inputText" />
```

File: src/controllers/customerCommandController.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    { commandName: string; commandKey: string; commandAction: Function }[]
+  > = ref([]);
+
+  function getCustomerCommands(
+    textInput: string,
+    commandsInput: {
+      commandName: string;
+      commandKey: string;
+      commandAction: Function;
+    }[]
```
Comment (nntai) [inline src/controllers/customerCommandController.ts:21]: Should define `Command` type and export it => other component or end-user could use it

File: src/places/firstPlace.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+export const isText = (arr: string[], text: string) => {
```
Comment (nntai) [inline src/places/firstPlace.ts:1]:
```suggestion
export const isCorrectKey = (arr: string[], text: string) => {
```

File: src/controllers/keysController.test.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  const customerModalController = modalController();
+
+  const keysInputController = keysController(customerModalController.onModalChange, "Control+k", [{commandName: "place", commandKey: "m", commandAction: () => {}}]);
+
+
+  it("keys controller", () => {
+
+
+    keysInputController.addKey("m");
+    expect((-1 !== keysInputController.keys.value.indexOf("m")) === true);
```
Comment (nntai) [inline src/controllers/keysController.test.js:23]:
```suggestion
    expect((keysInputController.keys.value.indexOf("m") !== -1) === true);
```

**Review by datngovan — APPROVED** [review]

**Review by lhtris — APPROVED** [review]

**Review by haonguyen3817884 — COMMENTED** [review]
approve

**Review by haonguyen3817884 — APPROVED** [review]

**Pull request merged**